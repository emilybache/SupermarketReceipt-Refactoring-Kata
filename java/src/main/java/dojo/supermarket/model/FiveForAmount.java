package dojo.supermarket.model;

import java.util.Optional;

public class FiveForAmount implements Offer {
    public static final int FIVE_ITEMS = 5;
    private double amount;

    public FiveForAmount(final double amount) {
        this.amount = amount;
    }

    @Override
    public Optional<Discount> calculateDiscount(final double unitPrice, final Product p, final double quantity) {
        int quantityAsInt = (int) quantity;
        if (quantityAsInt < FIVE_ITEMS) {
            return Optional.empty();
        }

        int numberOfXs = quantityAsInt / FIVE_ITEMS;
        double discountTotal = unitPrice * quantity - (amount * numberOfXs + quantityAsInt % 5 * unitPrice);
        return Optional.of(new Discount(p, FIVE_ITEMS + " for " + amount, discountTotal));
    }
}
