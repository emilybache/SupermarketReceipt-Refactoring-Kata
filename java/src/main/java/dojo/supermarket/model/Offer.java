package dojo.supermarket.model;

public abstract class Offer {
    SpecialOfferType offerType;
    double argument;

    Offer(SpecialOfferType offerType, double argument) {
        this.offerType = offerType;
        this.argument = argument;
    }

    public static Offer createOffer(SpecialOfferType offerType, Product product, double argument) {
        switch (offerType) {
            case ThreeForTwo: {
                return new ThreeForTwo(product, argument);
            }
            case TwoForAmount: {
                return new TwoForAmount(product, argument);
            }
            case FiveForAmount: {
                return new FiveForAmount(product, argument);
            }
            case TenPercentDiscount: {
                return new TenPercentDiscount(product, argument);
            }
            default:
                throw new IllegalArgumentException("Invalid offerType: " + offerType.name());
        }
    }

    protected abstract void doHandleOffers(final Receipt receipt, final SupermarketCatalog catalog, final Product p, final ShoppingCart shoppingCart);
}
