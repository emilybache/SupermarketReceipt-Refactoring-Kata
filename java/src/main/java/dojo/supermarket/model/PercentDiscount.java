package dojo.supermarket.model;

import java.util.Optional;

public class PercentDiscount implements Offer {
    private final double percentage;

    public PercentDiscount(final double percentage) {
        this.percentage = percentage;
    }

    @Override
    public Optional<Discount> calculateDiscount(final double unitPrice, final Product p, final double quantity) {
        return Optional.of(new Discount(p, percentage + "% off", quantity * unitPrice * percentage / 100.0));
    }
}
