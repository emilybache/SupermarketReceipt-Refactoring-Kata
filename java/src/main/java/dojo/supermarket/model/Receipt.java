package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Receipt {
    private List<ReceiptItem> items = new ArrayList<>();
    private List<Discount> discounts = new ArrayList<>();

    public Double getTotalPrice() {
        double total = 0.0;
        for (ReceiptItem item : this.items) {
            total += item.getTotalPrice();
        }
        for (Discount discount : this.discounts) {
            total -= discount.getDiscountAmount();
        }
        return total;
    }

    public void addProduct(Product p, double quantity, double price, double totalPrice) {
        this.items.add(new ReceiptItem(p, quantity, price, totalPrice));
    }

    public List<ReceiptItem> getItems() {
        return new ArrayList<>(this.items);
    }

    public void addDiscount(Discount discount) {
        this.discounts.add(discount);
    }

    public List<Discount> getDiscounts() {
        return discounts;
    }
}
