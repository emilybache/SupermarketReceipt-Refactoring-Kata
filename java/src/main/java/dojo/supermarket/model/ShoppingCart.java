package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ShoppingCart {

    private final List<ProductQuantity> items = new ArrayList<>();
    Map<Product, Double> productQuantities = new HashMap<>();


    List<ProductQuantity> getItems() {
        return new ArrayList<>(items);
    }

    void addItem(Product product) {
        this.addItemQuantity(product, 1.0);
    }

    Map<Product, Double> productQuantities() {
        return productQuantities;
    }


    public void addItemQuantity(Product product, double quantity) {
        items.add(new ProductQuantity(product, quantity));
        if (productQuantities.containsKey(product)) {
            productQuantities.put(product, productQuantities.get(product) + quantity);
        } else {
            productQuantities.put(product, quantity);
        }
    }

    void handleOffers(Receipt receipt, Map<Product, Offer> offers, SupermarketCatalog catalog) {
        for (Product p: productQuantities().keySet()) {
            double quantity = productQuantities.get(p);
            if (offers.containsKey(p)) {
                Offer offer = offers.get(p);
                double unitPrice = catalog.getUnitPrice(p);
                int quantityAsInt = (int) quantity;
                Discount discount = null;
                int x = 1;
                if (offer.offerType == SpecialOfferType.ThreeForTwo) {
                    x = 3;

                } else if (offer.offerType == SpecialOfferType.TwoForAmount) {
                    x = 2;
                    if (quantityAsInt >= 2) {
                        int intDivision = quantityAsInt / x;
                        double pricePerUnit = offer.argument * intDivision;
                        double theTotal = (quantityAsInt % 2) * unitPrice;
                        double total = pricePerUnit + theTotal;
                        double discountN = unitPrice * quantity - total;
                        discount = new Discount(p, "2 for " + offer.argument, discountN);
                    }

                } if (offer.offerType == SpecialOfferType.FiveForAmount) {
                    x = 5;
                }
                int numberOfXs = quantityAsInt / x;
                if (offer.offerType == SpecialOfferType.ThreeForTwo && quantityAsInt > 2) {
                    double discountAmount = quantity * unitPrice - ((numberOfXs * 2 * unitPrice) + quantityAsInt % 3 * unitPrice);
                    discount = new Discount(p, "3 for 2", discountAmount);
                }
                if (offer.offerType == SpecialOfferType.TenPercentDiscount) {
                    discount = new Discount(p, offer.argument + "% off", quantity * unitPrice * offer.argument / 100.0);
                }
                if (offer.offerType == SpecialOfferType.FiveForAmount && quantityAsInt >= 5) {
                    double discountTotal = unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % 5 * unitPrice);
                    discount = new Discount(p, x + " for " + offer.argument, discountTotal);
                }
                if (discount != null)
                    receipt.addDiscount(discount);
            }

        }
    }
}
