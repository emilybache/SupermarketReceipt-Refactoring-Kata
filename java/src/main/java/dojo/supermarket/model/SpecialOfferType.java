package dojo.supermarket.model;

public enum SpecialOfferType {
    ThreeForTwo, PercentDiscount, TwoForAmount, FiveForAmount;
}
