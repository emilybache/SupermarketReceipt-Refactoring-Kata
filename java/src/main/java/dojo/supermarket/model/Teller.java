package dojo.supermarket.model;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Teller {

    private final SupermarketCatalog catalog;
    private Map<Product, Offer> offers = new HashMap<>();

    public Teller(SupermarketCatalog catalog) {
        this.catalog = catalog;
    }

    public void addSpecialOffer(SpecialOfferType offerType, Product product, double argument) {
        this.offers.put(product, Offer.createOffer(offerType, product, argument));
    }

    public Receipt checksOutArticlesFrom(ShoppingCart theCart) {
        Receipt receipt = new Receipt();
        List<ProductQuantity> productQuantities = theCart.getItems();
        for (ProductQuantity pq: productQuantities) {
            Product p = pq.getProduct();
            double quantity = pq.getQuantity();
            double unitPrice = this.catalog.getUnitPrice(p);
            double price = quantity * unitPrice;
            receipt.addProduct(p, quantity, unitPrice, price);
        }
        theCart.handleOffers(receipt, this.offers, this.catalog);

        return receipt;
    }

}
