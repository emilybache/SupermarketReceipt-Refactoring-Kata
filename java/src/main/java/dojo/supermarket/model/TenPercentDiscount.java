package dojo.supermarket.model;

public class TenPercentDiscount extends Offer {
    public TenPercentDiscount(final Product product, final double argument) {
        super(SpecialOfferType.TenPercentDiscount, argument);
    }

    @Override
    protected void doHandleOffers(final Receipt receipt, final SupermarketCatalog catalog, final Product p, final ShoppingCart shoppingCart) {
        double quantity = shoppingCart.productQuantities.get(p);
        double unitPrice = catalog.getUnitPrice(p);
        Discount discount = new Discount(p, argument + "% off", quantity * unitPrice * argument / 100.0);
        receipt.addDiscount(discount);

    }
}
