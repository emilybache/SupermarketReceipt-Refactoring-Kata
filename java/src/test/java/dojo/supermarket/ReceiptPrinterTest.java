package dojo.supermarket;

import dojo.supermarket.model.*;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

public class ReceiptPrinterTest {

    Product toothbrush = new Product("toothbrush", ProductUnit.Each);
    Product apples = new Product("apples", ProductUnit.Kilo);
    Receipt receipt = new Receipt();

    @Test
    public void oneLineItem() {
        receipt.addProduct(toothbrush, 1, 0.99, 0.99);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void quantityTwo() {
        receipt.addProduct(toothbrush, 2, 0.99,0.99 * 2);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void looseWeight() {
        receipt.addProduct(apples, 2.3, 1.99,1.99 * 2.3);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void total() {

        receipt.addProduct(toothbrush, 1, 0.99, 2*0.99);
        receipt.addProduct(apples, 0.75, 1.99, 1.99*0.75);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void discounts() {
        receipt.addDiscount(new Discount(apples, "3 for 2", 0.99));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void printWholeReceipt() {
        receipt.addProduct(toothbrush, 1, 0.99, 0.99);
        receipt.addProduct(toothbrush, 2, 0.99, 2*0.99);
        receipt.addProduct(apples, 0.75, 1.99, 1.99*0.75);
        receipt.addDiscount(new Discount(toothbrush, "3 for 2", 0.99));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

}
