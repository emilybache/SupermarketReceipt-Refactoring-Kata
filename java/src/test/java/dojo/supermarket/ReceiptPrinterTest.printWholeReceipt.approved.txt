toothbrush                          0.99
toothbrush                          1.98
  0.99 * 2
apples                              1.49
  1.99 * 0.750
3 for 2(toothbrush)                -0.99

Total:                              3.47